File: src/main/java/io/scriptor/elf/SymbolTable.java
package io.scriptor.elf;

import org.jetbrains.annotations.NotNull;

import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;

public final class SymbolTable implements Iterable<Symbol> {

    private static final String NULL = "(null)";

    private final TreeMap<Long, String> map = new TreeMap<>();

    public @NotNull String resolve(final long addr) {
        if (map.isEmpty()) {
            return NULL;
        }

        final var entry = map.floorEntry(addr);
        if (entry == null) {
            return NULL;
        }

        if (entry.getKey() == addr) {
            return entry.getValue();
        }

        return "%s+0x%x".formatted(entry.getValue(), addr - entry.getKey());
    }

    public void put(final long addr, final @NotNull String name) {
        map.put(addr, name);
    }

    public long get(final @NotNull String name) {
        return map.entrySet()
                  .stream()
                  .filter(entry -> entry.getValue().equals(name))
                  .map(Map.Entry::getKey)
                  .mapToLong(Long::longValue)
                  .findAny()
                  .orElse(~0L);
    }

    @Override
    public @NotNull Iterator<Symbol> iterator() {
        return map.entrySet()
                  .stream()
                  .map(entry -> new Symbol(entry.getKey(), entry.getValue()))
                  .iterator();
    }
}

File: src/main/java/io/scriptor/elf/SymbolTable.kt
package io.scriptor.elf

import io.scriptor.util.Log.format
import java.util.*

class SymbolTable : Iterable<Symbol> {
    private val map = TreeMap<ULong, String>()

    fun resolve(addr: ULong): String {
        if (map.isEmpty()) {
            return NULL
        }

        val entry = map.floorEntry(addr)
        if (entry == null) {
            return NULL
        }

        if (entry.key == addr) {
            return entry.value
        }

        return format("%s+0x%x", entry.value, addr - entry.key)
    }

    fun put(addr: ULong, name: String) {
        map[addr] = name
    }

    override fun iterator(): MutableIterator<Symbol> {
        return map.entries
            .stream()
            .map { entry -> Symbol(entry.key, entry.value) }
            .iterator()
    }

    companion object {
        private const val NULL = "(null)"
    }
}
